### PR TITLE
Fix catastrophic cancellation in ELU/SELU by replacing exp(x)-1 with expm1(x)

### LINKED ISSUE #116943: `tf.nn.selu` and `tf.nn.elu` return 0 for small negative inputs (catastrophic cancellation in `exp(x) - 1`)
### Issue type

Bug

### Have you reproduced the bug with TensorFlow Nightly?

Yes

### Source

source

### TensorFlow version

TensorFlow 2.22.0-dev20260421 (nightly)

### Custom code

Yes

### OS platform and distribution

_No response_

### Mobile device

_No response_

### Python version

_No response_

### Bazel version

_No response_

### GCC/compiler version

_No response_

### CUDA/cuDNN version

_No response_

### GPU model and memory

_No response_

### Current behavior?

# `tf.nn.selu` and `tf.nn.elu` return 0 for small negative inputs (catastrophic cancellation in `exp(x) - 1`)

## Summary

`tf.nn.selu` and `tf.nn.elu` compute `exp(x) - 1` via separate `exp` and subtraction. For small negative x, `exp(x)` rounds to `1.0` in float32, producing `exp(x) - 1 = 0` (catastrophic cancellation). The result is 100% wrong for |x| ≤ 1e-8, and already 1.3% wrong at x = -1e-6.

XLA (`jit_compile=True`) is not affected — it uses `expm1` internally and returns correct results.

## Scale of the problem

SELU(x) for x < 0 is `scale × alpha × (exp(x) - 1)`. For small negative x, the correct value is approximately `1.7581 × x`.

| x | SELU expected | SELU eager | Relative error |
|---|----------|---------------|----------------|
| -1e-04 | -1.758e-4 | -1.758e-4 | 0.02% |
| -1e-06 | -1.758e-6 | -1.781e-6 | **1.33%** |
| -1e-07 | -1.758e-7 | -2.096e-7 | **19.21%** |
| -1e-08 | -1.758e-8 | **0.0** | **100%** |
| -1e-10 | -1.758e-10 | **0.0** | **100%** |

ELU has the same pattern (with alpha = 1):

| x | ELU expected | ELU eager | Relative error |
|---|----------|---------------|----------------|
| -1e-06 | -1.000e-6 | -1.013e-6 | **1.33%** |
| -1e-07 | -1.000e-7 | -1.192e-7 | **19.21%** |
| -1e-08 | -1.000e-8 | **0.0** | **100%** |

### Gradient is also affected (SELU)

Because the forward pass returns 0 for small negative x, the backward pass takes the wrong branch (`scale` instead of `scale × alpha × exp(x)`), producing ~40% gradient error:

```
x = -1e-07: selu_grad = 1.7581 (correct)
x = -1e-08: selu_grad = 1.0507 (wrong — should be 1.7581, ~40% error)
```

## Root cause

```
exp(-1e-8)     = 1.0       (in float32 — rounds to 1)
exp(-1e-8) - 1 = 0.0       (catastrophic cancellation)
expm1(-1e-8)   = -1e-8     (correct, no cancellation)
```

The fix is to replace `exp(x) - 1` with `expm1(x)` in the Eigen SELU/ELU kernels. XLA already does this, which is why `jit_compile=True` produces correct results.

## Environment

- TensorFlow 2.22.0-dev20260421 (nightly)
- CPU and GPU eager kernels both affected
- XLA is NOT affected

### Standalone code to reproduce the issue

```shell
## Reproduction


import tensorflow as tf

x = tf.constant([[-1e-10]], dtype=tf.float32)

print(tf.nn.selu(x).numpy())   # [[0.]]  — should be [[-1.758e-10]]
print(tf.nn.elu(x).numpy())    # [[0.]]  — should be [[-1e-10]]

# XLA gives the correct result:
print(tf.function(tf.nn.selu, jit_compile=True)(x).numpy())  # [[-1.758e-10]]
print(tf.function(tf.nn.elu, jit_compile=True)(x).numpy())   # [[-1e-10]]
```

### Relevant log output

```shell

```

### PR DESCRIPTION
**Pull request by Mahajan-Sachin**
### Description

Fix catastrophic floating-point cancellation in `tf.nn.elu` and `tf.nn.selu` eager kernels.

**Fixes #116943**

### Problem

The ELU and SELU Eigen kernels compute `exp(x) - 1` using separate `exp()` and subtraction operations. For small negative `x` in float32, `exp(x)` rounds to `1.0`, producing `exp(x) - 1 = 0.0` (100% error for |x| <= 1e-8).

XLA (`jit_compile=True`) is **not affected** — it already uses `expm1` internally.

### Bug Reproduction (BEFORE fix)

image

### After Fix Verification
image
image

### Fix

Replace `exp(x) - 1` with `expm1(x)` in 3 locations:

1. **`relu_op_functor.h`** — `Elu` functor: `features.exp() - features.constant(1)` → `features.expm1()`
2. **`relu_op_functor.h`** — `Selu` functor: same pattern
3. **`fused_eigen_output_kernels.h`** — Fused `Elu` activation: same pattern

`expm1` is already available in Eigen (`scalar_expm1_op`) and used by TensorFlow's `Expm1` op. This aligns eager kernel behavior with XLA.

### CODE REVIEW

**Comment by Mahajan-Sachin** [comment]
Hi @cantonios, gentle follow-up — this PR has been in the review queue 
for about 3 weeks now with all CI checks passing. The corresponding 
issue (#116943) has already been verified and labeled "awaiting PR merge" 
by @Kayyuri.

Would you be able to take a look when you get a chance? Happy to make 
any changes if needed. Thank you!

**Comment by djhdusjbd-eng** [comment]
ok thanks will wait

**Comment by keerthanakadiri** [comment]
Hi @Mahajan-Sachin , Can you kindky sign the CLA? Thank you!

**Comment by Mahajan-Sachin** [comment]
Hi @keerthanakadiri, I have already signed the Google Individual CLA linked to sachinmahajan903@gmail.com and GitHub username mahajan-sachin. Screenshots attached for reference. Please let me know if anything else is needed!
 
image

**Comment by Mahajan-Sachin** [comment]
Hi @keerthanakadiri @Kayyuri, I noticed a merge conflict had arisen with upstream master in relu_op_functor.h. I've resolved it by combining the upstream NaN handling (isnan() check) with my expm1() fix — both changes are now present together. Please let me know if the resolution looks correct!

**Comment by Mahajan-Sachin** [comment]
Hi @keerthanakadiri, just wanted to flag that the Gemini Code Assist review is complete and shows no issues — the changes are marked as correct and following best practices. Please let me know if there's anything else needed from my side!

**Comment by Mahajan-Sachin** [comment]
Hi @keerthanakadiri, this PR has been waiting for review for about 
3 months now. All 7 CI checks are passing, CLA is signed, and the 
Gemini Code Assist review found no issues.

There's also a pending "GitHub Actions Scan" check that seems to need 
maintainer approval to run.

@cantonios was assigned as reviewer but hasn't been able to review yet. 
Would it be possible to assign an alternate reviewer or approve the 
pending workflow? The corresponding issue (#116943) was already verified 
and labeled "awaiting PR merge" by @Kayyuri.

Thank you!

**Review by dmiltr3 — CHANGES_REQUESTED** [review]
We have reviewed your changes and they look mostly good. However, there are a few minor things to address before we can merge this:

### 1. Fix Compilation Error
In `tensorflow/core/kernels/relu_op_functor.h`, inside the `Selu` functor, after replacing `features.exp() - features.constant(one)` with `features.expm1()`, the variable `one` becomes unused. This causes compilation failures on systems with strict compiler flags (unused variable errors).

Please remove the unused variable:
```cpp
// In Selu functor
const auto scale = static_cast<T>(1.0507009873554804934193349852946);
const auto scale_alpha = static_cast<T>(1.7580993408473768599402175208123);
// const auto one = static_cast<T>(1); // <-- Remove this line
const auto zero = static_cast<T>(0);
```

### 2. Add Unit Tests
To verify this fix and prevent future regressions, we should add unit tests covering this catastrophic cancellation case.

Could you please add the following unit test cases to verify that small negative inputs (like `-1e-8`) do not yield exactly zero?

For `EluTest` in `tensorflow/python/kernel_tests/nn_ops/relu_op_test.py`:
```python
  def testSmallNegativeInputs(self):
    for t in [np.float32]:
      with ops.device("/device:CPU:0"):
        x = np.array([-1e-8]).astype(t)
        y = nn_ops.elu(x)
        self.assertNotEqual(self.evaluate(y)[0], 0.0)
```

For `SeluTest` in the same file:
```python
  def testSmallNegativeInputs(self):
    for t in [np.float32]:
      with ops.device("/device:CPU:0"):
        x = np.array([-1e-8]).astype(t)
        y = nn_ops.selu(x)
        self.assertNotEqual(self.evaluate(y)[0], 0.0)
```

### 3. Update Reference Implementations in Tests
Also in `relu_op_test.py`, please update the Numpy reference implementations `_npElu` and `_npSelu` to use `np.expm1` instead of `np.exp() - 1`. This ensures consistency between the reference and the implementation for these small values.

For `_npElu`:
```python
  def _npElu(self, np_features):
    return np.where(np_features < 0, np.expm1(np_features), np_features)
```

For `_npSelu`:
```python
  def _npSelu(self, np_features):
    scale = 1.0507009873554804934193349852946
    scale_alpha = 1.7580993408473768599402175208123
    return np.where(
        np_features < 0,
        scale_alpha * np.expm1(np_features),
        scale * np_features,
    )
```

---

### (Optional) FYI
There might be orther locations in the codebase that could benefit from similar improvements, such as `MklEluGradOp` in `tensorflow/core/kernels/mkl/mkl_relu_op.cc`. While not required for this PR, it might be worth checking out for consistency.

**Review by dmiltr3 — DISMISSED** [review]
We have simulated your changes locally and run the internal presubmit tests.

## Status
- **Local Application:** Success
- **Presubmit Tests (TAP):** **Passed** (0 failures).

## Details
Your changes to use `expm1` in `Elu` and `Selu` have been correctly applied to:
- `tensorflow/core/kernels/fused_eigen_output_kernels.h`
- `tensorflow/core/kernels/relu_op_functor.h`

This optimization is highly valued as it avoids catastrophic cancellation for small inputs.

**Comment by Mahajan-Sachin** [comment]
Hi @dmiltr3, thanks for the thorough review! All requested changes 
have been addressed:

1. The unused `one` variable was already removed during the earlier 
   merge conflict resolution.
2. Added `testSmallNegativeInputs` regression tests for both EluTest 
   and SeluTest.
3. Updated `_npElu` and `_npSelu` reference implementations to use 
   `np.expm1`.

Please take another look. Thank you!

**Comment by Mahajan-Sachin** [comment]
Hi @dmiltr3, all requested changes have been pushed and all 9 CI 
checks are passing. Could you please re-review when you get a chance? 
Thank you!

**Review by dmiltr3 — APPROVED** [review]

**Comment by Mahajan-Sachin** [comment]
Hi @dmiltr3, all 12 CI checks have passed. Let me know if anything 
else is needed. Thank you!

**Comment by Mahajan-Sachin** [comment]
any follow up?

**Comment by nithyak0204** [comment]
Hi @Mahajan-Sachin, Sorry for the delay, this PR is waiting for internal approvals. We are working on it. Thank you!

**Comment by Mahajan-Sachin** [comment]
No problem sir, i just wanted best for community like you, take your time. I will start more prs soon in tf, exam ended. Now i am free. .
